### DISCUSSION TITLE
WebdriverIO - Why isn't waitUntil() working?

### DISCUSSION THREAD

**Discussion by RevNight** [webdriverio/webdriverio#6746] [Q&A] [answered]
I am trying to get a WaitUntil() example to work and I don't know why it isn't. I am using Chrome, WebdriverIO, VSC IDE, javascript, node pm etc. Basically, I want to continuously monitor this Date/Time site and when the hour:min:sec listed is equal to the value I specify, console logs a 'Time equals'. Pretty basic, this is just a proof of concept. Yet even I never can get this log.

I am pretty sure that I am declared the CurrentTime and DesiredTime correctly. To verify this, I wrote the boolean CurrentTime > DesiredTime. So if the CurrentTime on the site is 5:00:00pm, and my DesiredTime is 4:40:10 pm, boolean shows 'true', which is correct. Similiarly, if the CurrentTime is 5:00:00pm and the DesiredTime is 5:10:10pm, the boolean shows 'false', which is correct. This proves to me that the values are being considered to be numbers, and thus can be compared to, instead of just text.

So it should be simple then to write a WaitUntil for these two times to be equal. I have used =, ==, and === and all fail. I have used return in the condition too.

Also, as soon as the script runs, I immediately see the timeout message ('they dont equal'). How is this appearing so quickly?

I think the problem lies somewhere in syntax, but I can't figure it out.

To reproduce:
1. Load the website and look at the time
2. Change the DesiredTime value to be ~45 seconds into the future. Keep in mind the timeout
3. Run script. Wait a bit and notice how real world (CurrentTime) time value is eventually displayed, yet console doesn't display 'Time Equals'.

```js
it('WaitUntil', ()=> {
    browser.url('https://www.timeanddate.com/worldclock/usa/washington-dc')
    let CurrentTime = $("(//span[contains(@id, 'ct')])[1]").getText()
    console.log(CurrentTime)
    let DesiredTime = '4:20:40 pm'
    console.log(Boolean(CurrentTime > DesiredTime))
    browser.waitUntil(function(){
        CurrentTime == DesiredTime},
        {timeout: 30000}, console.log('they dont equal')
    )
    console.log('Time Equals')
```

**Comment by christian-bromann** [comment]
@RevNight you need to return a value in the `waitUntil function, e.g.:

```js
    browser.waitUntil(() => {
        return CurrentTime === DesiredTime
    }, {timeout: 30000})
```

**Reply by RevNight** [reply]
Hi @christian-bromann ,

That didn't work. I ran the below code and the failure of 'they dont equal' immediately was written. Also, when the times did equal each other, the flag of 'Times Equal' wasn't thrown.


```js
        browser.waitUntil(() => {return CurrentTime === DesiredTime},
                {timeout: 30000}, console.log('they dont equal')                 
        )
```

**Reply by christian-bromann** [reply]
Your code is incorrect which is why I posted a suggestion in mine

**Reply by christian-bromann** [reply]
Also have you tried log both values to see if they ever have the chance to be equal?

**Comment by erwinheitzman** [comment, marked as answer]
It should be like the following:
```js
const configObject = {
  timeout: 30000,
  interval: 200,
  timeoutMsg: 'the date/time did not match'
};

browser.waitUntil(() => {
  return $("(//span[contains(@id, 'ct')])[1]").getText() === '4:20:40 pm';
}, configObject);
```

reference: https://webdriver.io/docs/api/browser/waitUntil/

**Reply by RevNight** [reply]
@erwinheitzman wins the day! While I am thankful that you showed me the right answer, I am pretty frustrated that I can't use variables in WaitUntil. I took your exact same syntax, put in the declared Variable names, and it didn't work. Is there a reason that WaitUntil can't work with variables?

**Reply by RevNight** [reply]
ok so you can use variables, but you have to declare the one from the website INSIDE the WaitUntil(). This is kinda dumb, I don't see why you can't declare it in the above section. Very well!

```js
    it('WaitUntil', ()=> {
        browser.url('https://www.timeanddate.com/worldclock/usa/washington-dc')
        let DesiredTime = '11:29:50 am'
        const configObject = {
            timeout: 30000,
            interval: 200,
            timeoutMsg: 'the date/time did not match'
          };
          browser.waitUntil(() => {
            let CurrentTime = $("(//span[contains(@id, 'ct')])[1]").getText()
            return CurrentTime === DesiredTime;
          }, configObject);
        console.log('Time Equals')   

    })
```

**Reply by christian-bromann** [reply]
To clarify: the function you pass in `waitUntil` is called repeatedly until you return `true` which is why you need this lined (`let CurrentTime = $("(//span[contains(@id, 'ct')])[1]").getText()` in there because if you define this value before then it would never be updated.
